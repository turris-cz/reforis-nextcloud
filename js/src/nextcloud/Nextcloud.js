/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";

import {
    useAlert,
    SpinnerElement,
    formFieldsSize,
    useAPIGet,
    useWSForisModule,
    withErrorMessage,
    withSpinnerOnSending,
} from "foris";
import PropTypes from "prop-types";

import ConfigurationForm from "./ConfigurationForm";
import API_URLs from "../API";

Nextcloud.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Nextcloud({ ws }) {
    const [nextcloud, getNextcloud] = useAPIGet(API_URLs.nextcloud);

    useEffect(() => {
        getNextcloud();
    }, [getNextcloud]);

    return (
        <>
            <h1>{_("Nextcloud")}</h1>
            <p>
                {_(
                    "Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home."
                )}
            </p>
            <ConfigurationWithErrorAndSpinner
                ws={ws}
                apiState={nextcloud.state}
                nextcloud={nextcloud.data}
                getNextcloud={getNextcloud}
            />
        </>
    );
}

const ConfigurationWithErrorAndSpinner = withErrorMessage(
    withSpinnerOnSending(Configuration)
);

Configuration.propTypes = {
    ws: PropTypes.object.isRequired,
    nextcloud: PropTypes.object.isRequired,
    getNextcloud: PropTypes.func.isRequired,
};

function Configuration({ ws, nextcloud, getNextcloud }) {
    const [setAlert] = useAlert();
    const [isConfiguring, setIsConfiguring] = useState(false);
    const [configurationStateWS] = useWSForisModule(
        ws,
        "nextcloud",
        "state_change"
    );

    useEffect(() => {
        if (!configurationStateWS) return;

        if (configurationStateWS.configuration === "failed") {
            setAlert(configurationStateWS.msg);
            getNextcloud();
        }
        if (configurationStateWS.configuration === "completed") {
            getNextcloud();
        }
    }, [configurationStateWS, getNextcloud, setAlert]);

    const { nextcloud_configuring, nextcloud_configured } = nextcloud;

    let componentContent;
    if (nextcloud_configuring || isConfiguring) {
        componentContent = (
            <div className="d-flex flex-row justify-content-center text-muted">
                <SpinnerElement small>
                    <p className="ml-1">{_("Configuring Nextcloudâ€¦")}</p>
                </SpinnerElement>
            </div>
        );
    } else if (nextcloud_configured) {
        componentContent = (
            <div className="text-muted text-center">
                <p className="mb-2">{_("Congratulations!ðŸŽ‰")}</p>
                <p
                    dangerouslySetInnerHTML={{
                        __html: _(
                            'You have configured your Nextcloud now you can visit the <a href="/nextcloud" target="_blank" rel="noopener noreferrer">site<sup><i class="fas fa-external-link-alt fa-sm ml-1"></i></sup></a>.'
                        ),
                    }}
                />
            </div>
        );
    } else {
        componentContent = (
            <ConfigurationForm setIsConfiguring={setIsConfiguring} />
        );
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Nextcloud Configuration")}</h2>
            <p>{_("Create an admininstrator account for Nextcloud.")}</p>
            {componentContent}
        </div>
    );
}
