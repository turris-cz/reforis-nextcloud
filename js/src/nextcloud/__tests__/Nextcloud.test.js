/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";

import {
    act,
    render,
    wait,
    getByText,
    getByLabelText,
    fireEvent,
} from "foris/testUtils/customTestRender";
import { WebSockets } from "foris";
import { mockJSONError } from "foris/testUtils/network";

import Nextcloud from "../Nextcloud";
import {
    nextcloud_configured,
    nextcloud_configuring,
} from "./__fixtures__/nextcloud";
import nextcloud_default from "./__fixtures__/nextcloud";

describe("<Nextcloud />", () => {
    let webSockets;
    let container;

    beforeEach(() => {
        webSockets = new WebSockets();
        ({ container } = render(<Nextcloud ws={webSockets} />));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render error", async () => {
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/nextcloud/api/nextcloud",
            expect.anything()
        );
        mockJSONError();
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should render Nextcloud", async () => {
        mockAxios.mockResponse({ data: nextcloud_default });
        await wait(() =>
            expect(getByText(container, "Nextcloud Configuration")).toBeTruthy()
        );
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should render Nextcloud configuring", async () => {
        mockAxios.mockResponse({ data: nextcloud_configuring });
        await wait(() =>
            expect(getByText(container, "Configuring Nextcloudâ€¦")).toBeTruthy()
        );
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should render Nextcloud configured", async () => {
        mockAxios.mockResponse({ data: nextcloud_configured });
        await wait(() =>
            expect(getByText(container, "Congratulations!ðŸŽ‰")).toBeTruthy()
        );
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should configure Nextcloud", async () => {
        mockAxios.mockResponse({ data: nextcloud_default });
        await wait(() =>
            expect(getByText(container, "Nextcloud Configuration")).toBeTruthy()
        );

        const username = "some_username";
        const password = "some_password";
        const usernameInput = getByLabelText(container, "Username");
        const passwordInput = getByLabelText(container, "Password");
        const submitButton = getByText(container, "Save");

        fireEvent.change(usernameInput, { target: { value: username } });
        fireEvent.change(passwordInput, { target: { value: password } });
        fireEvent.click(submitButton);

        expect(mockAxios.post).toBeCalledWith(
            "/reforis/nextcloud/api/nextcloud",
            {
                credentials: {
                    login: username,
                    password: password,
                },
            },
            expect.anything()
        );
        mockAxios.mockResponse({ data: nextcloud_configuring });

        await wait(() =>
            expect(getByText(container, "Configuring Nextcloudâ€¦")).toBeTruthy()
        );

        act(() =>
            webSockets.dispatch({
                module: "nextcloud",
                action: "state_change",
                data: { configuration: "completed" },
            })
        );

        expect(mockAxios.get).toHaveBeenNthCalledWith(
            2,
            "/reforis/nextcloud/api/nextcloud",
            expect.anything()
        );
        mockAxios.mockResponse({ data: nextcloud_configured });

        await wait(() =>
            expect(getByText(container, "Congratulations!ðŸŽ‰")).toBeTruthy()
        );
    });
});
