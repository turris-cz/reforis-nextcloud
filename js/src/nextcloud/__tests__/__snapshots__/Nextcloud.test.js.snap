// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nextcloud /> should render Nextcloud 1`] = `
<div>
  <h1>
    Nextcloud
  </h1>
  <p>
    Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Nextcloud Configuration
    </h2>
    <p>
      Create an administrator account for Nextcloud.
    </p>
    <form>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="1"
        >
          Username
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control is-invalid"
            id="1"
            type="text"
            value=""
          />
        </div>
        <div
          class="invalid-feedback"
        >
          Username cannot be empty.
        </div>
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="2"
        >
          Password
        </label>
        <div
          class="input-group"
        >
          <input
            autocomplete="new-password"
            class="form-control is-invalid"
            id="2"
            type="password"
            value=""
          />
          <button
            class="input-group-text"
            type="button"
          >
            <i
              class="fa fa-eye"
            />
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Password cannot be empty.
        </div>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          disabled=""
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<Nextcloud /> should render Nextcloud configured 1`] = `
<div>
  <h1>
    Nextcloud
  </h1>
  <p>
    Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Nextcloud Configuration
    </h2>
    <p>
      Create an administrator account for Nextcloud.
    </p>
    <div
      class="text-muted text-center"
    >
      <p
        class="mb-2"
      >
        Congratulations!ðŸŽ‰
      </p>
      <p
        class="mb-0"
      >
        You have configured your Nextcloud now you can visit the 
        <a
          href="/nextcloud"
          rel="noopener noreferrer"
          target="_blank"
        >
          site
          <sup>
            <i
              class="fas fa-external-link-alt fa-sm ms-1"
            />
          </sup>
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;

exports[`<Nextcloud /> should render Nextcloud configuring 1`] = `
<div>
  <h1>
    Nextcloud
  </h1>
  <p>
    Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Nextcloud Configuration
    </h2>
    <p>
      Create an administrator account for Nextcloud.
    </p>
    <div
      class="d-flex flex-row justify-content-center align-items-center text-muted"
    >
      <div
        class="spinner-border spinner-border-sm"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
      <div
        class="spinner-text"
      >
        <p
          class="ms-1 mb-0"
        >
          Configuring Nextcloudâ€¦
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Nextcloud /> should render error 1`] = `
<div>
  <h1>
    Nextcloud
  </h1>
  <p>
    Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home.
  </p>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<Nextcloud /> should render spinner 1`] = `
<div>
  <h1>
    Nextcloud
  </h1>
  <p>
    Nextcloud puts your data at your fingertips, under your control. Store your documents, calendar, contacts and photos on a server at home.
  </p>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
